27_tictactoe/main.py

```diff
@@ -0,0 +1,46 @@
+import re
+import lib.board as board
+from lib.player import Player
+
+players = [None,None]
+i = 0
+
+while True:
+
+    # Get players' info
+    while i < 2:
+        try:
+            players[i] = Player(name   = raw_input("Player %d - Name:   " %(i)),
+                                marker = raw_input("Player %d - Marker: " %(i)))
+            i += 1
+        except Exception as e:
+            print "Error occured while setting player %d\nError message: %s" %(i, e.message)
+
+
+    # Init game
+    tto = board.Game()
+    tto.start(players[0], players[1])
+
+    # Play game
+    while (tto.winner is None) and (not tto.tie()):
+        current_play = raw_input("%s's play (enter row,col): " % tto.current_player.name)
+        m = re.search(r'^\s*(\w)[,;\s\.]*(\w)\s*$', current_play)
+        try:
+            tto.play(int(m.group(1)),int(m.group(2)))
+            print tto
+        except board.InvalidMove as e:
+            print e.message
+        except Exception as e:
+            print e.message
+
+    # Print results
+    if tto.tie(): print "Game tied!!"
+    else: print "Winner is %s - well done!!\n" %(tto.winner.name)
+    print "------- GAME OVER --------\n"
+
+    # Continue?
+    if raw_input("Would you like to play again (y/n)? ").lower() == 'n': 
+        print "\nGood bye... Hope you enjoyed!!\n"
+        break
+    if raw_input("Same players (y/n)? ").lower() == 'n': 
+        i = 0
```

Comment:
I still believe that there was an issue with the original test since taking (col,row) instead of (row,col) should have the same effect as transposing the board which should make no difference on the result.
The sequence below generates:
X O X
O X O
X X O

Which means X wins.

From: ynonp [mailto:notifications@github.com]
Sent: Friday, July 29, 2016 9:31 AM
To: ynonp/python-examples-verint-2016-07
Cc: Naor, Nir; Author
Subject: Re: [ynonp/python-examples-verint-2016-07] ToCode Solution: תרגיל מסכם מחלקות (#221)


In 27_tictactoe/main.py<https://github.com/ynonp/python-examples-verint-2016-07/pull/221#discussion_r72746918>:

> +        except board.InvalidMove as e:

> +            print e.message

> +        except Exception as e:

> +            print e.message

> +

> +    # Print results

> +    if tto.tie(): print "Game tied!!"

> +    else: print "Winner is %s - well done!!\n" %(tto.winner.name)

> +    print "------- GAME OVER --------\n"

> +

> +    # Continue?

> +    if raw_input("Would you like to play again (y/n)? ").lower() == 'n':

> +        print "\nGood bye... Hope you enjoyed!!\n"

> +        break

> +    if raw_input("Same players (y/n)? ").lower() == 'n':

> +        i = 0

Problem was not in Player class but in tie detection. In the test play method is expected to get (col, row), and in your game it takes (row, col). That's why the play sequence in the test does not product a tie and the test fails.

This is the sequence that according to the test was to product a tie, but in your game produces victory for the first player:

28 self.g.play(0,0)
29 self.g.play(0,1)
30 self.g.play(0,2)
31 self.g.play(1,1)
32 self.g.play(1,0)
33 self.g.play(1,2)
34 self.g.play(2,1)
35 self.g.play(2,2)
36 self.g.play(2,0)

That's not too bad, just solving in a different way.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ynonp/python-examples-verint-2016-07/pull/221/files/e7ea17bf98533b25fcd24aa58240127dd7112296#r72746918>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATWlup1dXKoSP88vcSzq26HSyknq4r1fks5qaZ4ngaJpZM4JXyZi>.

-- 
This electronic message may contain proprietary and confidential 
information of Verint Systems Inc., its affiliates and/or subsidiaries.
The information is intended to be for the use of the individual(s) or
entity(ies) named above. If you are not the intended recipient (or 
authorized to receive this e-mail for the intended recipient), you may not 
use, copy, disclose or distribute to anyone this message or any information 
contained in this message. If you have received this electronic message in 
error, please notify us by replying to this e-mail.